### DISCUSSION TITLE
Error create database, migration

### DISCUSSION THREAD

**Discussion by exptjs** [mikro-orm/mikro-orm#5491] [Q&A]
Can you tell me why the database and migration are not being created?

```
// package.json
dependencies": {
    "@mikro-orm/cli": "^6.2.2",
    "@mikro-orm/core": "^6.2.2",
    "@mikro-orm/postgresql": "^6.2.2",
    "express": "^4.19.2",
    "pg": "^8.11.5",
    "reflect-metadata": "^0.2.2",
  },
  "devDependencies": {
    "@types/express": "^4.17.21",
    "@types/node": "^20.11.30",
    "ts-node": "^10.9.2",
    "typescript": "^5.4.3"
  },
  "mikro-orm": {
    "useTsNode": true,
    "configPaths": [
      "./src/mikro-orm.config.ts",
      "./dist/mikro-orm.config.js"
    ]
  }
```

```
// mikro-orm.config.ts
import {Options, PostgreSqlDriver} from '@mikro-orm/postgresql';

const configDb: Options = {
  host: 'localhost',
  port: 5432,
  driver: PostgreSqlDriver,
  schema: 'public'
  dbName: 'test_mikro'
  user: 'USER',
  password: 'PASSWORD',
  entities: ['./dist/entities/*.js'],
  entitiesTs: ['./src/entities/*.ts'],
  ignoreUndefinedInQuery: true,
  discovery: {
    warnWhenNoEntities: false,
    requireEntitiesArray: false,
    disableDynamicFileAccess: false,
  },
  debug: true
};
```

```
mikro-orm debug

Current MikroORM CLI configuration
 - dependencies:    
   - mikro-orm 6.2.2
   - node 20.11.1                                                               
   - knex 3.1.0                                                                
   - pg 8.11.5                                                                 
   - typescript 5.4.3                                                          
 - package.json found                                                          
 - ts-node enabled                                                             
 - searched config paths:                                                      
   - xxx/src/mikro-orm.config.ts (found)     
   - xxx/dist/mikro-orm.config.js (not found)
   - xxx/mikro-orm.config.ts (not found)     
   - xxx/mikro-orm.config.js (not found)
 - configuration found
 - database connection failed (���� ������ "test_mikro" �� ����������)
 - will use `entities` array (contains 0 references and 1 paths)
   - xxx/dist/entities/*.js (not found)
 - could use `entitiesTs` array (contains 0 references and 1 paths)      
   - xxx/src/entities/*.ts (not found)
 ```

If create the database manually: `- database connection successful`

```
mikro-orm migration:create --name table_a

[info] MikroORM version: 6.2.2
[discovery] ORM entity discovery started, using ReflectMetadataProvider
[discovery] - processing 0 files
[discovery] - entity discovery finished, found 0 entities, took 12 ms
[query] select 1 from pg_database where datname = 'test_mikro' [took 43 ms]
mikro-orm database:create

Create your database if it does not exist

error: ���� ������ "test_mikro" �� ����������
    at Parser.parseErrorMessage (xxx\node_modules\pg-protocol\src\parser.ts:369:69)
    at Parser.handlePacket (xxx\node_modules\pg-protocol\src\parser.ts:188:21)
    at Parser.parse (xxx\node_modules\pg-protocol\src\parser.ts:103:30)
    at Socket.<anonymous> (xxx\node_modules\pg-protocol\src\index.ts:7:48)
    at Socket.emit (node:events:518:28)
    at Socket.emit (node:domain:488:12)
    at addChunk (node:internal/streams/readable:559:12)
    at readableAddChunkPushByteMode (node:internal/streams/readable:510:3)
    at Socket.Readable.push (node:internal/streams/readable:390:5)
    at TCP.onStreamRead (node:internal/stream_base_commons:190:23) {
  length: 97,
  severity: '�����',
  code: '3D000',
  detail: undefined,
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'postinit.c',
  line: '887',
  routine: 'InitPostgres'
}
```
If create the database manually: 
```
mikro-orm migration:create --name table_a

[info] MikroORM version: 6.2.2
[discovery] ORM entity discovery started, using ReflectMetadataProvider
[discovery] - processing 0 files
[discovery] - entity discovery finished, found 0 entities, took 9 ms
[query] select 1 from pg_database where datname = 'test_mikro' [took 46 ms, 1 result]
[info] MikroORM successfully connected to database test_mikro on postgresql://USER:*****@localhost:5432
mikro-orm migration:create

Create new migration with current schema diff

Error: Migrator extension not registered.
    at MikroORM.getMigrator (xxx\node_modules\@mikro-orm\core\MikroORM.js:228:15)
    at Function.handleMigrationCommand (xxx\node_modules\@mikro-orm\cli\commands\MigrationCommandFactory.js:84:30)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
```

**Comment by B4nan** [comment]
> error: ���� ������ "test_mikro" �� ����������

No idea what this means, sounds like you use some system language that uses special characters. I would first focus on getting the actual error message instead of this. I would guess it's about permissions, are you sure your user can create new databases? The `code: '3D000',` part should give you some clues.

> Error: Migrator extension not registered.

This means exactly what it says, you need to register migrator extension to be able to work with it.

https://mikro-orm.io/docs/configuration#extensions

**Comment by exptjs** [comment]
> I would guess it's about permissions, are you sure your user can create new databases? The `code: '3D000',` part should give you some clues.

Yes, there is permission to create a database. If run `npx sequelize-cli model:generate --name table_a`, it works

> https://mikro-orm.io/docs/configuration#extensions

Thank you, it helped me, I didn't think it was mandatory.

Executing `npx mikro-orm migration:create --initial` created the `mikro_orm_migrations` table.

Do I understand correctly that by executing `npx micro-orm migration:create --name table_a` a migration file for the table `table_a` should be created?
By executing `npx mikro-orm migration:create --name table_a` in the `migrations` directory, the migration file is not created.

In `sequelize`, executing `npx sequelize-cli model:generate --name table_a` creates a migration file and a model. How can I create a model in `micro-orm`, or is there no model here, but something else?

**Reply by B4nan** [reply]
> Do I understand correctly that by executing npx micro-orm migration:create --name table_a a migration file for the table table_a should be created?

No, a migration with a name `table_a` (or better say name suffix) will be generated, based on your entity metadata. In your example there are apparently no entities, so it won't generate anything - you need to define the entities first, or use `--blank` argument to create an empty migration you can fill in yourself.

> By executing npx mikro-orm migration:create --name table_a in the `migrations' directory, the migration file is not created.

You always need to execute this from root of your project, running it elsewhere is wrong.

> In sequelize, executing npx sequelize-cli model:generate --name table_a

Generating a model and a migration are two very different things.

**Reply by exptjs** [reply]
Thanks, now it's clear how migrations and entities work.
Only database creation does not work, although the user has access rights to create it.

**Comment by exptjs** [comment]
Please look at the [example](https://github.com/exptjs/test-mikro-orm).

Maybe i missed something because I didn't find a complete basic example.

`npx mikro-orm database:create` causes an error `code: '3D000'` even if specify a database user `postgres`.

Entities and migrations are created, and the database is updated successfully.
